Add OVA build scripts and utilities

frapposelli opened the pull request:
This is a very big PR, I separated the two vendor commits for ease of review.

There are a few things missing like the version is hardcoded and there are a few rough edges but it's in working conditions and ready to be iterated on.

/cc @akutz @sflxn @figo

dougm commented on installer/build/bootable/root/etc/inputrc:
```diff
@@ -0,0 +1,42 @@
+"\e[A":history-search-backward
+"\e[B":history-search-forward
+
+$if mode=emacs
```
:+1:

dougm commented on installer/build/build-ova.sh:
```diff
@@ -0,0 +1,70 @@
+EOF
+
+echo -e "--------------------------------------------------
+buidling ova with env...\n
```
typo

dougm commented on installer/build/scripts/provisioners/provision_govc.sh:
```diff
@@ -0,0 +1,28 @@
+${GOVC_SHA256}
+EOF
+
+curl -L"#" -o govc_linux_amd64.gz https://github.com/vmware/govmomi/releases/download/v0.17.1/govc_linux_amd64.gz
```
not 0.19?

figo commented on installer/build/container/push.sh:
```diff
@@ -0,0 +1,31 @@
+
+OVA_REV=$(git rev-parse --verify --short=8 HEAD)
+IMAGE="cluster-api-ova-build"
+# TODO(frapposelli): find a better home for the build container
```
👍  agree that a official account is needed

frapposelli replied:
moved to `gcr.io/cnx-cluster-api/cluster-api-ova-build`

akutz commented:
Re: signed binaries -- we can always place them in our VMC AWS S3 bucket. The thing is, if you commit them you can *never* remove them without rewriting them out of the repo's history. That will change every commit ID after the binaries were committed. I favor storing them in S3.

sidharthsurana commented:
> Re: signed binaries -- we can always place them in our VMC AWS S3 bucket. The thing is, if you commit them you can _never_ remove them without rewriting them out of the repo's history. That will change every commit ID after the binaries were committed. I favor storing them in S3.

+2 this idea of not storing binaries directly into the git repo. One additional tool that we can explore is https://github.com/jedbrown/git-fat This is an add-on tool for git. We use this tool internally to store binaries (reference to) in our git repo and it works pretty much as expected. The benefit of this tool is you still get to maintain the entire history of changes in the git repository without burdening the git repo with actual binary files. However, I am not sure if it supports S3 directly.

akutz commented:
@sidharthsurana 

I don't even think a tool is necessary. As @frapposelli said, the binaries have their own versions. We just place them in s3 with a versioned file name reflecting the correct version, and we're good to go.

frapposelli commented:
Ok, let’s move those off to a GS or S3 bucket :+1:

@akutz can you follow up with me offline with access to a bucket? I’ll amend the PR after the upload is done.

sidharthsurana commented:
@akutz  Using the tool just gives you more native feeling that these binaries are checked in to your git repository and are versioned by git.
But for this case either way works.

akutz commented:
FWIW @frapposelli, I have no major objection other than I know from very recent experience that it's not fun to remove them. As this is a LTS project, I imagine we don't want to rewrite the git history at a later date.

zjs commented on installer/build/bootable/config/builder.ovf:
```diff
@@ -0,0 +1,147 @@
+            FullVersion is a descriptive version string
+            including, for example, alpha or beta designations
+            and other release criteria.
+        -->
```
Is this comment accurate?

frapposelli replied:
Yes, it describes how `--version--` is substituted with the proper version during the build process

zjs replied:
On the subsequent lines, it seems like `FullVersion` is blank?

zjs commented on installer/ovatools/ova-ui/main.go:
```diff
@@ -0,0 +1,270 @@
+	si := &syscall.Sysinfo_t{}
+	err := syscall.Sysinfo(si)
+	if err != nil {
+		panic("Austin, we have a problem... syscall.Sysinfo:" + err.Error())
```
This may be a confusing panic message to some.

zjs commented on installer/build/build.sh:
```diff
@@ -0,0 +1,71 @@
+ROOT_INSTALLER_WORK_DIR="${ROOT_WORK_DIR}/installer"
+
+# TAG=${DRONE_TAG:-$(git describe --abbrev=0 --tags)} # e.g. `v0.9.0`
+TAG="v0.0.1"
```
Is hard-coding the version intentional? (The above reference to `DRONE_TAG` should be removed, at least.)

frapposelli replied:
Yes, it is intentional, I will be adding a comment to better explain it in code

zjs commented on installer/build/build.sh:
```diff
@@ -0,0 +1,71 @@
+  export BUILD_NUMBER=${BUILD_NUMBER}
+  export DRONE_BUILD_NUMBER=${DRONE_BUILD_NUMBER}
+  export DRONE_BUILD_EVENT=${DRONE_BUILD_EVENT}
+  export DRONE_DEPLOY_TO=${DRONE_DEPLOY_TO}
```
These references to `DRONE` should be removed.

zjs commented on installer/build/build-ova.sh:
```diff
@@ -0,0 +1,70 @@
+function setenv() {
+  tmpvar=$1
+  fallback=$2
+  # if a cli argument was specified, it's a local file, gcloud bucket upload, or revision number.
```
Is this accurate?

zjs commented on installer/build/build-ova.sh:
```diff
@@ -0,0 +1,70 @@
+done
+
+# set Kubernetes Version
+setenv KUBERNETES "$KUBERNETES_DEFAULT_VERSION"
```
Is this now equivalent to `KUBERNETES=${KUBERNETES:-$KUBERNETES_DEFAULT_VERSION}`?

frapposelli commented on installer/build/bootable/build-app.sh:
```diff
@@ -0,0 +1,110 @@
+sed -i 's/umask 027/umask 022/' /etc/profile
+
+log3 "setting root password"
+echo 'root:Vmw@re!23' | chpasswd
```
this password is overridden during first boot as specifying a root password is required to deploy the OVA. SSH is set to never accept passwords and only work with keys.

frapposelli commented on installer/Makefile:
```diff
@@ -0,0 +1,119 @@
+	@echo building ova-ui
+	@GOARCH=amd64 GOOS=linux $(TIME) $(GO) build $(RACE) -ldflags "$(LDFLAGS)" -o ./$@ ./$(dir $<)
+
+$(toolbox): $(call godeps,toolbox/*.go)
```
I'm a fan of `toolbox` for simple use cases like this one, it also gives us the ability to extend its functionality (like [this](
https://github.com/vmware/govmomi/tree/master/toolbox#httproundtripper)) to accommodate more complex automation with Cluster API

akutz replied:
Hi @frapposelli / @zjs,

akutz replied:
Hi @frapposelli / @zjs,

FWIW, yakity uses `open-vm-tools`, and I often have issues with them taking a while to come online in terms of visibility to the vCenter. That said, I use `open-vm-tools` because it's one less thing...

dougm replied:
Note that toolbox does support all of the guest operations.  It is possible to only expose a subset (as VIC does) or disable all (as this PR does).  The main issue is that toolbox only has a hook for in-guest authentication, but leaves it to the consumer to implement. 
https://github.com/vmware/govmomi/tree/master/toolbox#guestfilemanager-object
https://github.com/vmware/govmomi/tree/master/toolbox#guestprocessmanager

Regarding the visibility in vCenter, open-vm-tools takes ~30 sec (last I checked) to initially report IP addresses, toolbox is around 1 sec or less.

sflxn commented on installer/Gopkg.toml:
```diff
@@ -0,0 +1,36 @@
+#  name = "github.com/x/y"
+#  version = "2.4.0"
+
+required = ["github.com/golang/dep", "github.com/GoASTScanner/gas", "github.com/go-yaml/yaml", "github.com/golang/lint/golint"]
```
We're pulling in kubernetes-sigs/yaml in the provider project.  It's based on go-yaml.  Kubernetes projects seem to like using forked versions (into the k8s.io or sigs.k8s.io repos) of outside projects.

frapposelli replied:
I'm fine with pulling the same as long as they're compatible

frapposelli commented:
@zjs thanks a lot for reviewing 👍 I addressed most of your comments in a separate commit

The plan is to squash all the review commits into one once it's lgtm'd

sflxn commented on installer/build/container/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+    gawk util-linux which docker gptfdisk e2fsprogs grub2-efi efibootmgr parted dosfstools xz docker; \
+    curl -L'#' -k https://dl.google.com/dl/cloudsdk/channels/rapid/downloads/google-cloud-sdk-200.0.0-linux-x86_64.tar.gz  | tar xzf - -C /usr/local; \
+    mkdir -p /root/.gsutil/; \
+    /usr/local/google-cloud-sdk/install.sh --quiet; \
```
Why do we need google-cloud-sdk?  It's quite large and takes awhile to install.

frapposelli replied:
removed

sflxn commented on installer/build/container/push.sh:
```diff
@@ -0,0 +1,30 @@
+
+OVA_REV=$(git rev-parse --verify --short=8 HEAD)
+IMAGE="cluster-api-ova-build"
+REPO="docker.io/frapposelli/"
```
We now have a repo for the provider container if you want to use the same repo.  gcr.io/cnx-cluster-api (part of VMW's gcr account).  You just need the GCR key file, which I can provide.

frapposelli replied:
let's follow up offline with that and I'll amend once it's pushed

frapposelli replied:
pushed and amended

frapposelli commented on installer/Makefile:
```diff
@@ -0,0 +1,119 @@
+# Copyright 2018 VMware, Inc. All Rights Reserved.
```
updated the headers, double checking with legal but it should be good to go

frapposelli commented:
@akutz @figo @dougm @zjs @sflxn 

Thanks for the thorough review guys, I should have adressed all the major concerns here, last two were moving the build container to a more permanent location (`gcr.io` under the `cnx-cluster-api` project) and removing the boot/EFI binaries from the repo to a google storage bucket under the same project.

Please take a second look at it, I'll squash everything to the second commit once it LGTY.

dougm commented:
@frapposelli sorry I should have added a `/hold`.  Wasn't expecting it to merge since there wasn't an `/approve`, but I guess that's implicit since you are an approver.

frapposelli commented:
No worries, *I* should’ve put a /hold on it :sad:
